server/controllers/admin/entities/index.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * HTTP END POINT
+ * API for the entities/ http end point
+ */
+const util = require('../../../lib/util');
+const db = require('../../../lib/db');
+const types = require('./types');
+
+exports.types = types;
+exports.list = list;
+exports.details = details;
+exports.update = update;
+exports.remove = remove;
+exports.create = create;
+
+function list(req, res, next) {
+  const query = `
+    SELECT 
+      BUID(e.uuid) AS uuid, e.display_name, e.gender, e.email, e.phone, e.address, 
+      e.reference, et.id AS entity_type_id, et.label, et.translation_key
+    FROM entity e
+    JOIN entity_type et ON et.id = e.entity_type_id
+  `;
+  db.exec(query)
+    .then(rows => res.status(200).json(rows))
+    .catch(next)
+    .done();
+}
+
+function details(req, res, next) {
+  const buid = db.bid(req.params.uuid);
+  fetchEntity(buid)
+    .then(entity => res.status(200).json(entity))
+    .catch(next)
+    .done();
+}
+
+/**
+ * PUT /entities/:uuid
+ * return 200 status code instead of 204 because an object is returned for need of tests
+ */
+function update(req, res, next) {
+  const query = `
+    UPDATE entity SET ? WHERE uuid = ?;
+  `;
+
+  const params = req.body;
+  const buid = db.bid(req.params.uuid);
+
+  if (params.uuid) {
+    delete params.uuid;
+  }
+
+  db.exec(query, [params, buid])
+    .then(() => fetchEntity(buid))
+    .then(entity => res.status(200).json(entity))
+    .catch(next)
+    .done();
+}
+
+/**
+ * DELETE /entities/:uuid
+ * return 200 status code instead of 204 because an empty array is returned for need of tests
+ */
+function remove(req, res, next) {
+  const query = `
+    DELETE FROM entity WHERE uuid = ?;
+  `;
+  const buid = db.bid(req.params.uuid);
+  db.exec(query, [buid])
+    .then(() => fetchEntity(buid, true))
+    .then((rows) => res.status(200).json(rows))
+    .catch(next)
+    .done();
+}
+
+function create(req, res, next) {
+  const query = `
+    INSERT INTO entity SET ?;
+  `;
+  const params = req.body;
+  const identifier = params.uuid || util.uuid();
+  params.uuid = db.bid(identifier);
+  db.exec(query, [params])
+    .then(() => res.status(201).json({ uuid : identifier }))
+    .catch(next)
+    .done();
+}
+
+/**
+ * @function fetchEntity
+ * @param {object} uuid a binary uuid
+ * @param {boolean} allowEmpty allow to return an empty array instead of an error
+ */
+function fetchEntity(uuid, allowEmpty) {
+  const query = `
+    SELECT 
+      BUID(e.uuid) AS uuid, e.display_name, e.gender, e.email, e.phone, e.address, 
+      e.reference, et.id AS entity_type_id, et.label, et.translation_key
+    FROM entity e
+    JOIN entity_type et ON et.id = e.entity_type_id
+    WHERE uuid = ?;
+  `;
+  return allowEmpty ? db.exec(query, [uuid]).then(rows => rows) : db.one(query, [uuid]);
```

Comment:
`.then(rows => rows)` is not necessary.

It looks like these are two fundamentally different queries.  Why do we need to allowEmpty?